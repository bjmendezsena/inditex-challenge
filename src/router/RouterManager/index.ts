export * from './RouteName';
export * from './RoutePath';
