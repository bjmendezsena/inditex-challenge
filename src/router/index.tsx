import { RouterProvider, createBrowserRouter } from "react-router-dom";
import routes from "./Routes";

export const router = createBrowserRouter(routes);

const AppRouter = () => <RouterProvider router={router} />;

export default AppRouter;
