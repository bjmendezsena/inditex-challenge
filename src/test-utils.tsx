import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { Router } from "react-router-dom";
import { createMemoryHistory, MemoryHistory } from "history";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { theme } from "./theme";

export const testQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      refetchOnMount: true,
      cacheTime: 0,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: process.env.NODE_ENV === "test" ? () => {} : console.error,
  },
});

export const queryWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

const DefaultTheme = extendTheme(theme);

const memoryHistory = createMemoryHistory();

interface RenderOptions {
  history?: MemoryHistory;
}

export const render = (
  ui: JSX.Element,
  { history: customHistory }: RenderOptions = {}
): RenderResult => {
  const history = customHistory || memoryHistory;
  const wrapper = ({ children }: { children?: React.ReactNode }) => (
    <QueryClientProvider client={testQueryClient}>
      <ChakraProvider theme={DefaultTheme}>
        <Router location={history.location} navigator={history}>
          {children}
        </Router>
      </ChakraProvider>
    </QueryClientProvider>
  );

  return testingLibraryRender(ui, { wrapper });
};
