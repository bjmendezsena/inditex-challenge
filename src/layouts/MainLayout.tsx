import { useEffect, useState } from "react";
import { Link as RRDLink, useLocation } from "react-router-dom";
import { Flex } from "@chakra-ui/react";
import { Outlet } from "react-router-dom";
import { Header, Link, NavigationIndicator } from "../components";
import { RoutePath } from "../router/RouterManager";

export const MainLayout = () => {
  const location = useLocation();
  const [isNavigating, setIsNavigating] = useState(false);

  useEffect(() => {
    setIsNavigating(false);
    return () => {
      setIsNavigating(true);
    };
  }, [location]);

  return (
    <Flex w='100vw' h='100vh' flexDirection='column' bg='white'>
      <Header
        leftContent={
          <Link as={RRDLink} to={RoutePath.PodcastList}>
            Podcaster
          </Link>
        }
        rightContent={<NavigationIndicator isNavigating={isNavigating} />}
      />
      <Outlet />
    </Flex>
  );
};
