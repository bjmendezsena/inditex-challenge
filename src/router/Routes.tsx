import { RouteObject, Navigate } from "react-router-dom";
import * as Views from "../views";
import { MainLayout, PodcastLayout } from "../layouts";
import { RoutePath } from "./RouterManager";

const routes: RouteObject[] = [
  {
    element: <MainLayout />,
    children: [
      {
        path: RoutePath.PodcastList,
        element: <Views.PodcastListPage />,
      },
      {
        element: <PodcastLayout />,
        children: [
          {
            path: RoutePath.PodcastDetails,
            element: <Views.PodcastDetails />,
          },
          {
            path: RoutePath.EpisodeDetails,
            element: <Views.EpisodeDetails />,
          },
        ],
      },
      {
        path: "*",
        element: <Navigate to={RoutePath.PodcastList} replace />,
      },
    ],
  },
];

export default routes;
