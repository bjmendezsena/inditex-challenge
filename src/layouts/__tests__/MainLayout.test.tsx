import { screen } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { matchPath } from "react-router-dom";
import { render } from "../../test-utils";
import { MainLayout } from "../MainLayout";
import { RouterManager, RouteName } from "../../router";

describe(`<${MainLayout.name}/>`, () => {
  const history = createMemoryHistory();
  const factoryComponent = () =>
    render(<MainLayout />, {
      history,
    });

  it("should render without crashing", () => {
    const { container } = factoryComponent();
    expect(container).toMatchSnapshot();
  });

  it("Should navigate to home page when click on podcast link", () => {
    factoryComponent();
    const logo = screen.getByText("Podcaster");
    expect(logo).toBeInTheDocument();
    expect(logo).toHaveAttribute("href", "/");
    logo.click();
    expect(
      matchPath(
        RouterManager.getUrl(RouteName.PodcastList),
        history.location.pathname
      )
    ).toBeTruthy();
  });
});
