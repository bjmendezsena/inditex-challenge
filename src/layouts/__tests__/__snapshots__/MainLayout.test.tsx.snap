// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MainLayout/> should render without crashing 1`] = `
<div>
  <div
    class="css-tw0lc1"
  >
    <div
      class="css-1gzl4dy"
      data-testid="header"
    >
      <div
        class="css-1k3povc"
      >
        <a
          class="chakra-link css-2otp49"
          data-testid="link"
          href="/"
          tabindex="0"
        >
          Podcaster
        </a>
        <div
          class="css-15vddy2"
        />
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
